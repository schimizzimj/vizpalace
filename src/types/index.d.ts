export interface DataPoint<T = string | number | Date> {
  x: T; // x-axis value, can be a category, numeric value or date
  y: number | null; // y-axis value
}

export interface ChartSeries<T = string | number | Date> {
  name: string; // name of the series
  color?: string; // optional color for the series
  values: DataPoint<T>[]; // array of data points
}

export type ChartData<T = string | number | Date> = ChartSeries<T>[];

interface XAxisOptions {
  enabled?: boolean;
}

interface YAxisOptions {
  enabled?: boolean;
}

interface AxisOptions {
  yAxis?: YAxisOptions;
  xAxis?: XAxisOptions;
}

export interface ChartOptions {
  width: number;
  height: number;
  margin?: {
    top: number;
    right: number;
    bottom: number;
    left: number;
  };
  animationDuration?: number;
}

export interface BarChartOptions extends ChartOptions, AxisOptions {
  displayType?: "stacked" | "grouped";
}
