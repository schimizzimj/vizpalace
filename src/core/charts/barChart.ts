import * as d3 from "d3";
import {
  initializeChartDimensions,
  createSVG,
  appendXAxis,
  appendYAxis,
} from "../base/chartUtils";
import { BarData, ChartOptions } from "../../types";

export function createBarChart(
  element: HTMLElement,
  data: BarData[],
  userOptions: Partial<ChartOptions> = {}
): void {
  // Merge default options with user options
  const options = initializeChartDimensions(userOptions);
  let { width, height, animationDuration } = options;
  const svg = createSVG(element, options);
  animationDuration = animationDuration ?? 1000;

  const xScale = d3
    .scaleBand()
    .domain(data.map((d) => d.label))
    .range([0, width])
    .padding(0.1);

  const yMax =
    d3.max(data, (d) =>
      d.value !== undefined && d.value !== null ? d.value : 0
    ) || 0;

  const yScale = d3.scaleLinear().domain([0, yMax]).range([height, 0]);

  const bars = svg
    .selectAll(".bar")
    .data(data)
    .enter()
    .append("rect")
    .attr("class", "bar")
    .attr("x", (d) => xScale(d.label) || 0)
    .attr("y", height)
    .attr("width", xScale.bandwidth())
    .attr("height", 0);

  bars
    .transition()
    .duration(animationDuration)
    .ease(d3.easeCubicInOut)
    .attr("y", (d) => yScale(d.value ?? 0))
    .attr("height", (d) => height - yScale(d.value ?? 0));

  appendXAxis(svg, xScale, height);
  appendYAxis(svg, yScale);
}
