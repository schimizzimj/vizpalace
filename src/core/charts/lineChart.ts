import * as d3 from "d3";
import {
  initializeChartDimensions,
  createSVG,
  appendXAxis,
  appendYAxis,
} from "../base/chartUtils";

export function createLineChart(
  element: HTMLElement,
  data: LineData[],
  userOptions: Partial<ChartOptions> = {}
): void {
  const options = initializeChartDimensions(userOptions);
  let { width, height, animationDuration } = options;

  const svg = createSVG(element, options);
  animationDuration = animationDuration ?? 0;

  const xScale = d3
    .scaleTime()
    .domain(d3.extent(data, (d) => d.date) as [Date, Date])
    .range([0, width]);

  const yMax =
    d3.max(data, (d) =>
      d.value !== undefined && d.value !== null ? d.value : 0
    ) || 0;

  const yScale = d3.scaleLinear().domain([0, yMax]).range([height, 0]);

  // Create the line generator
  const line = d3
    .line<LineData>()
    .x((d) => xScale(d.date))
    .y((d) => yScale(d.value ?? 0));

  // Draw the line
  const linePath = svg
    .append("path")
    .datum(data)
    .attr("class", "line")
    .attr("d", line)
    .attr("fill", "none")
    .attr("stroke", "steelblue")
    .attr("stroke-width", 1.5);

  if (linePath.node()?.getTotalLength) {
    const totalLength = linePath.node()?.getTotalLength() ?? 0;
    linePath
      .attr("stroke-dasharray", `${totalLength} ${totalLength}`)
      .attr("stroke-dashoffset", totalLength)
      .transition()
      .duration(animationDuration)
      .ease(d3.easeLinear)
      .attr("stroke-dashoffset", 0);
  }

  appendXAxis(svg, xScale, height);
  appendYAxis(svg, yScale);
}
