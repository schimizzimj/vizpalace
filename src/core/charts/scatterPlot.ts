import * as d3 from "d3";
import { ChartData, ScatterPlotOptions } from "../../types";
import { createSVG, initializeScatterPlotDimensions } from "../base/chartUtils";
import { appendXAxis, appendYAxis } from "../base/axes";
import { isScaleBand } from "../utils/utils";

export function createScatterPlot<T = number | string | Date>(
  element: HTMLElement,
  data: ChartData<T>,
  userOptions: Partial<ScatterPlotOptions> = {}
) {
  const options = initializeScatterPlotDimensions(userOptions);
  let { chartDimensions, animationDuration, xAxis, yAxis } = options;
  const { width, height } = chartDimensions;
  const svg = createSVG(element, options);
  const duration = animationDuration ?? 0;

  const allDataPoints = data.flatMap((series) => series.values);

  let xScale:
    | d3.ScaleBand<string>
    | d3.ScaleTime<number, number>
    | d3.ScaleLinear<number, number>;

  if (typeof allDataPoints[0].x === "string") {
    xScale = d3
      .scaleBand()
      .padding(0.1)
      .domain(allDataPoints.map((d) => d.x as string));
  } else if (allDataPoints[0].x instanceof Date) {
    xScale = d3
      .scaleTime()
      .domain(d3.extent(allDataPoints, (d) => d.x as Date) as [Date, Date]);
  } else {
    xScale = d3
      .scaleLinear()
      .domain(
        d3.extent(allDataPoints, (d) => d.x as number) as [number, number]
      );
  }

  xScale.range([0, width]);

  const yMin = d3.min(allDataPoints, (d) => d.y ?? 0) ?? 0;
  const yMax = d3.max(allDataPoints, (d) => d.y ?? 0) ?? 0;

  const yScale = d3.scaleLinear().domain([yMin, yMax]).range([height, 0]);

  let circlesWrapper: d3.Selection<SVGGElement, any, any, any> =
    svg.select(".circles-wrapper");

  if (circlesWrapper.empty()) {
    circlesWrapper = svg.append("g").attr("class", "circles-wrapper");
  }

  const circles = circlesWrapper
    .selectAll<SVGCircleElement, any>(".circle")
    .data(allDataPoints);

  circles.exit().remove();

  circles
    .enter()
    .append("circle")
    .attr("class", "circle")
    .attr("cx", (d) => {
      if (isScaleBand(xScale)) {
        return (xScale(d.x as string) ?? 0) + xScale.bandwidth() / 2;
      } else {
        return xScale(d.x as any) ?? 0;
      }
    })
    .attr("cy", (d) => yScale(d.y ?? 0) ?? 0)
    .attr("r", 5)
    .attr("fill", "steelblue")
    .merge(circles)
    .transition()
    .duration(duration)
    .attr("cx", (d) => {
      if (isScaleBand(xScale)) {
        return (xScale(d.x as string) ?? 0) + xScale.bandwidth() / 2;
      } else {
        return xScale(d.x as any) ?? 0;
      }
    })
    .attr("cy", (d) => yScale(d.y ?? 0) ?? 0);

  appendXAxis(svg, xScale, chartDimensions, xAxis ?? {});
  appendYAxis(svg, yScale, chartDimensions, yAxis ?? {});
}
