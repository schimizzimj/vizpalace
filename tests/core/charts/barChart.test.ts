import { assert, expect, test, describe, beforeEach, afterEach } from "vitest";
import { createBarChart } from "../../../src/core/charts/barChart";
import { ChartData } from "../../../src/types";

describe("barChart", () => {
  let container: HTMLDivElement;
  let data: ChartData;

  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
  });

  afterEach(() => {
    document.body.removeChild(container);
  });

  describe("single series", () => {
    beforeEach(() => {
      data = [
        {
          name: "Test",
          values: [
            { x: "foo", y: 10 },
            { x: "bar", y: 20 },
            { x: "baz", y: 30 },
          ],
        },
      ];
    });

    test("it should render a bar chart", () => {
      createBarChart(container, data, {});

      const svg = container.querySelector("svg");
      assert(svg);

      const bars = svg.querySelectorAll("rect");
      expect(bars.length).toBe(3);
    });

    test("it should use custom width and height", () => {
      const config = {
        width: 100,
        height: 100,
        margin: { top: 0, bottom: 0, left: 0, right: 0 },
      };
      createBarChart(container, data, config);

      const svg = container.querySelector("svg");
      assert(svg);
      expect(svg.getAttribute("width")).toBe("100");
      expect(svg.getAttribute("height")).toBe("100");
    });

    test("it should use custom margins", () => {
      const config = {
        width: 100,
        height: 100,
        margin: { top: 10, right: 10, bottom: 10, left: 10 },
      };
      createBarChart(container, data, config);

      const svg = container.querySelector("svg");
      assert(svg);
      expect(svg.getAttribute("width")).toBe(
        `${config.width + config.margin.left + config.margin.right}`
      );
      expect(svg.getAttribute("height")).toBe(
        `${config.height + config.margin.top + config.margin.bottom}`
      );
    });

    test("it should look visually correct", async () => {
      createBarChart(container, data, { animationDuration: 0 });
      await new Promise((resolve) => setTimeout(resolve, 1000));
      const result = container.innerHTML;
      expect(result).toMatchSnapshot();
    });
  });

  describe("multiple series", () => {
    beforeEach(() => {
      data = [
        {
          name: "Test 1",
          values: [
            { x: "foo", y: 10 },
            { x: "bar", y: 20 },
            { x: "baz", y: 30 },
          ],
        },
        {
          name: "Test 2",
          values: [
            { x: "foo", y: 20 },
            { x: "bar", y: 30 },
            { x: "baz", y: 40 },
          ],
        },
      ];
    });

    test("it should render a bar chart", () => {
      createBarChart(container, data, {});

      const svg = container.querySelector("svg");
      assert(svg);

      const bars = svg.querySelectorAll("rect");
      expect(bars.length).toBe(6);
    });

    test("it should look visually correct", async () => {
      createBarChart(container, data, { animationDuration: 0 });
      await new Promise((resolve) => setTimeout(resolve, 1000));
      const result = container.innerHTML;
      expect(result).toMatchSnapshot();
    });
  });
});
