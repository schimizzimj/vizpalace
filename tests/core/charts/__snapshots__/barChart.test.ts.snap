// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`barChart > multiple series > grouped > it should look visually correct 1`] = `"<svg width=\\"660\\" height=\\"450\\"><g transform=\\"translate(40, 20)\\"><rect class=\\"bar bar-0\\" x=\\"19.354838709677438\\" y=\\"300\\" width=\\"87.09677419354838\\" fill=\\"#1f77b4\\" height=\\"100\\"></rect><rect class=\\"bar bar-0\\" x=\\"212.90322580645162\\" y=\\"200\\" width=\\"87.09677419354838\\" fill=\\"#1f77b4\\" height=\\"200\\"></rect><rect class=\\"bar bar-0\\" x=\\"406.4516129032258\\" y=\\"100\\" width=\\"87.09677419354838\\" fill=\\"#1f77b4\\" height=\\"300\\"></rect><rect class=\\"bar bar-1\\" x=\\"106.45161290322582\\" y=\\"200\\" width=\\"87.09677419354838\\" fill=\\"#ff7f0e\\" height=\\"200\\"></rect><rect class=\\"bar bar-1\\" x=\\"300\\" y=\\"100\\" width=\\"87.09677419354838\\" fill=\\"#ff7f0e\\" height=\\"300\\"></rect><rect class=\\"bar bar-1\\" x=\\"493.54838709677415\\" y=\\"0\\" width=\\"87.09677419354838\\" fill=\\"#ff7f0e\\" height=\\"400\\"></rect><g class=\\"x-axis\\" transform=\\"translate(0, 400)\\" fill=\\"none\\" font-size=\\"10\\" font-family=\\"sans-serif\\" text-anchor=\\"middle\\"><path class=\\"domain\\" stroke=\\"currentColor\\" d=\\"M0.5,6V0.5H600.5V6\\"></path><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(106.45161290322582,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">foo</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(300,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">bar</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(493.54838709677415,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">baz</text></g></g><g class=\\"y-axis\\" fill=\\"none\\" font-size=\\"10\\" font-family=\\"sans-serif\\" text-anchor=\\"end\\"><path class=\\"domain\\" stroke=\\"currentColor\\" d=\\"M-6,400.5H0.5V0.5H-6\\"></path><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,400.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">0</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,350.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">5</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,300.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">10</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,250.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">15</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,200.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">20</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,150.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">25</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,100.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">30</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,50.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">35</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,0.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">40</text></g></g></g></svg>"`;

exports[`barChart > multiple series > stacked > it should look visually correct 1`] = `"<svg width=\\"660\\" height=\\"450\\"><g transform=\\"translate(40, 20)\\"><rect class=\\"bar bar-0\\" x=\\"19.354838709677438\\" y=\\"300\\" width=\\"87.09677419354838\\" fill=\\"#1f77b4\\" height=\\"100\\"></rect><rect class=\\"bar bar-0\\" x=\\"212.90322580645162\\" y=\\"200\\" width=\\"87.09677419354838\\" fill=\\"#1f77b4\\" height=\\"200\\"></rect><rect class=\\"bar bar-0\\" x=\\"406.4516129032258\\" y=\\"100\\" width=\\"87.09677419354838\\" fill=\\"#1f77b4\\" height=\\"300\\"></rect><rect class=\\"bar bar-1\\" x=\\"106.45161290322582\\" y=\\"200\\" width=\\"87.09677419354838\\" fill=\\"#ff7f0e\\" height=\\"200\\"></rect><rect class=\\"bar bar-1\\" x=\\"300\\" y=\\"100\\" width=\\"87.09677419354838\\" fill=\\"#ff7f0e\\" height=\\"300\\"></rect><rect class=\\"bar bar-1\\" x=\\"493.54838709677415\\" y=\\"0\\" width=\\"87.09677419354838\\" fill=\\"#ff7f0e\\" height=\\"400\\"></rect><g class=\\"x-axis\\" transform=\\"translate(0, 400)\\" fill=\\"none\\" font-size=\\"10\\" font-family=\\"sans-serif\\" text-anchor=\\"middle\\"><path class=\\"domain\\" stroke=\\"currentColor\\" d=\\"M0.5,6V0.5H600.5V6\\"></path><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(106.45161290322582,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">foo</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(300,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">bar</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(493.54838709677415,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">baz</text></g></g><g class=\\"y-axis\\" fill=\\"none\\" font-size=\\"10\\" font-family=\\"sans-serif\\" text-anchor=\\"end\\"><path class=\\"domain\\" stroke=\\"currentColor\\" d=\\"M-6,400.5H0.5V0.5H-6\\"></path><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,400.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">0</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,350.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">5</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,300.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">10</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,250.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">15</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,200.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">20</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,150.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">25</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,100.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">30</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,50.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">35</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,0.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">40</text></g></g></g></svg>"`;

exports[`barChart > single series > it should look visually correct 1`] = `"<svg width=\\"660\\" height=\\"450\\"><g transform=\\"translate(40, 20)\\"><rect class=\\"bar bar-0\\" x=\\"19.354838709677438\\" y=\\"266.6666666666667\\" width=\\"174.19354838709677\\" fill=\\"#1f77b4\\" height=\\"133.33333333333331\\"></rect><rect class=\\"bar bar-0\\" x=\\"212.90322580645162\\" y=\\"133.33333333333334\\" width=\\"174.19354838709677\\" fill=\\"#1f77b4\\" height=\\"266.66666666666663\\"></rect><rect class=\\"bar bar-0\\" x=\\"406.4516129032258\\" y=\\"0\\" width=\\"174.19354838709677\\" fill=\\"#1f77b4\\" height=\\"400\\"></rect><g class=\\"x-axis\\" transform=\\"translate(0, 400)\\" fill=\\"none\\" font-size=\\"10\\" font-family=\\"sans-serif\\" text-anchor=\\"middle\\"><path class=\\"domain\\" stroke=\\"currentColor\\" d=\\"M0.5,6V0.5H600.5V6\\"></path><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(106.45161290322582,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">foo</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(300,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">bar</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(493.54838709677415,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">baz</text></g></g><g class=\\"y-axis\\" fill=\\"none\\" font-size=\\"10\\" font-family=\\"sans-serif\\" text-anchor=\\"end\\"><path class=\\"domain\\" stroke=\\"currentColor\\" d=\\"M-6,400.5H0.5V0.5H-6\\"></path><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,400.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">0</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,373.8333333333333)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">2</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,347.1666666666667)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">4</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,320.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">6</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,293.83333333333337)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">8</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,267.1666666666667)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">10</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,240.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">12</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,213.83333333333334)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">14</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,187.16666666666666)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">16</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,160.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">18</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,133.83333333333334)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">20</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,107.16666666666669)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">22</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,80.49999999999999)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">24</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,53.83333333333332)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">26</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,27.16666666666666)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">28</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,0.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">30</text></g></g></g></svg>"`;
